File: 000Unordered_ Videos/WWDC 2011/WWDC_2011_SourceCode/Sample-Code-master/LazyTableImages/Classes/LazyTableAppDelegate.m
/*
     File: LazyTableAppDelegate.m 
 Abstract: Application delegate for the LazyTableImages sample.
 It also downloads in the background the "Top Paid iPhone Apps" RSS feed using NSURLConnection.
  
  Version: 1.3 
  
 Disclaimer: IMPORTANT:  This Apple software is supplied to you by Apple 
 Inc. ("Apple") in consideration of your agreement to the following 
 terms, and your use, installation, modification or redistribution of 
 this Apple software constitutes acceptance of these terms.  If you do 
 not agree with these terms, please do not use, install, modify or 
 redistribute this Apple software. 
  
 In consideration of your agreement to abide by the following terms, and 
 subject to these terms, Apple grants you a personal, non-exclusive 
 license, under Apple's copyrights in this original Apple software (the 
 "Apple Software"), to use, reproduce, modify and redistribute the Apple 
 Software, with or without modifications, in source and/or binary forms; 
 provided that if you redistribute the Apple Software in its entirety and 
 without modifications, you must retain this notice and the following 
 text and disclaimers in all such redistributions of the Apple Software. 
 Neither the name, trademarks, service marks or logos of Apple Inc. may 
 be used to endorse or promote products derived from the Apple Software 
 without specific prior written permission from Apple.  Except as 
 expressly stated in this notice, no other rights or licenses, express or 
 implied, are granted by Apple herein, including but not limited to any 
 patent rights that may be infringed by your derivative works or by other 
 works in which the Apple Software may be incorporated. 
  
 The Apple Software is provided by Apple on an "AS IS" basis.  APPLE 
 MAKES NO WARRANTIES, EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION 
 THE IMPLIED WARRANTIES OF NON-INFRINGEMENT, MERCHANTABILITY AND FITNESS 
 FOR A PARTICULAR PURPOSE, REGARDING THE APPLE SOFTWARE OR ITS USE AND 
 OPERATION ALONE OR IN COMBINATION WITH YOUR PRODUCTS. 
  
 IN NO EVENT SHALL APPLE BE LIABLE FOR ANY SPECIAL, INDIRECT, INCIDENTAL 
 OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
 SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
 INTERRUPTION) ARISING IN ANY WAY OUT OF THE USE, REPRODUCTION, 
 MODIFICATION AND/OR DISTRIBUTION OF THE APPLE SOFTWARE, HOWEVER CAUSED 
 AND WHETHER UNDER THEORY OF CONTRACT, TORT (INCLUDING NEGLIGENCE), 
 STRICT LIABILITY OR OTHERWISE, EVEN IF APPLE HAS BEEN ADVISED OF THE 
 POSSIBILITY OF SUCH DAMAGE. 
  
 Copyright (C) 2012 Apple Inc. All Rights Reserved. 
  
 */

#import "LazyTableAppDelegate.h"
#import "RootViewController.h"
#import "ParseOperation.h"

// This framework was imported so we could use the kCFURLErrorNotConnectedToInternet error code.
#import <CFNetwork/CFNetwork.h>


static NSString *const TopPaidAppsFeed =
	@"http://phobos.apple.com/WebObjects/MZStoreServices.woa/ws/RSS/toppaidapplications/limit=75/xml";


@interface LazyTableAppDelegate () 
{
    UIWindow				*window;
    UINavigationController	*navigationController;
	
    // this view controller hosts our table of top paid apps
    RootViewController      *rootViewController;
    
    // the list of apps shared with "RootViewController"
    NSMutableArray          *appRecords;
    
    // the queue to run our "ParseOperation"
    NSOperationQueue		*queue;
    
    // RSS feed network connection to the App Store
    NSURLConnection         *appListFeedConnection;
    NSMutableData           *appListData;
}
@end


@implementation LazyTableAppDelegate

@synthesize window, navigationController, appRecords, rootViewController, queue, appListFeedConnection, appListData;


#pragma mark -

// -------------------------------------------------------------------------------
//	applicationDidFinishLaunching:application
// -------------------------------------------------------------------------------
- (void)applicationDidFinishLaunching:(UIApplication *)application
{
    // Initializr the app window
    self.window = [[[UIWindow alloc] initWithFrame:[[UIScreen mainScreen] bounds]] autorelease];
    self.window.rootViewController = self.navigationController;
    [self.window makeKeyAndVisible];
    
    // Initialize the array of app records and pass a reference to that list to our root view controller
    self.appRecords = [NSMutableArray array];
    rootViewController.entries = self.appRecords;
    
    NSURLRequest *urlRequest = [NSURLRequest requestWithURL:[NSURL URLWithString:TopPaidAppsFeed]];
    self.appListFeedConnection = [[[NSURLConnection alloc] initWithRequest:urlRequest delegate:self] autorelease];
    
    // Test the validity of the connection object. The most likely reason for the connection object
    // to be nil is a malformed URL, which is a programmatic error easily detected during development
    // If the URL is more dynamic, then you should implement a more flexible validation technique, and
    // be able to both recover from errors and communicate problems to the user in an unobtrusive manner.
    //
    NSAssert(self.appListFeedConnection != nil, @"Failure to create URL connection.");
    
    // show in the status bar that network activity is starting
    [UIApplication sharedApplication].networkActivityIndicatorVisible = YES;
}

// -------------------------------------------------------------------------------
//	handleLoadedApps:notif
// -------------------------------------------------------------------------------
- (void)handleLoadedApps:(NSArray *)loadedApps
{
    [self.appRecords addObjectsFromArray:loadedApps];
    
    // tell our table view to reload its data, now that parsing has completed
    [rootViewController.tableView reloadData];
}

#pragma mark -
#pragma mark NSURLConnection delegate methods

// -------------------------------------------------------------------------------
//	handleError:error
// -------------------------------------------------------------------------------
- (void)handleError:(NSError *)error
{
    NSString *errorMessage = [error localizedDescription];
    UIAlertView *alertView = [[UIAlertView alloc] initWithTitle:@"Cannot Show Top Paid Apps"
														message:errorMessage
													   delegate:nil
											  cancelButtonTitle:@"OK"
											  otherButtonTitles:nil];
    [alertView show];
    [alertView release];
}

// The following are delegate methods for NSURLConnection. Similar to callback functions, this is how
// the connection object,  which is working in the background, can asynchronously communicate back to
// its delegate on the thread from which it was started - in this case, the main thread.
//

// -------------------------------------------------------------------------------
//	connection:didReceiveResponse:response
// -------------------------------------------------------------------------------
- (void)connection:(NSURLConnection *)connection didReceiveResponse:(NSURLResponse *)response
{
    self.appListData = [NSMutableData data];    // start off with new data
}

// -------------------------------------------------------------------------------
//	connection:didReceiveData:data
// -------------------------------------------------------------------------------
- (void)connection:(NSURLConnection *)connection didReceiveData:(NSData *)data
{
    [appListData appendData:data];  // append incoming data
}

// -------------------------------------------------------------------------------
//	connection:didFailWithError:error
// -------------------------------------------------------------------------------
- (void)connection:(NSURLConnection *)connection didFailWithError:(NSError *)error
{
    [UIApplication sharedApplication].networkActivityIndicatorVisible = NO;
    
    if ([error code] == kCFURLErrorNotConnectedToInternet)
	{
        // if we can identify the error, we can present a more precise message to the user.
        NSDictionary *userInfo = [NSDictionary dictionaryWithObject:@"No Connection Error"
															 forKey:NSLocalizedDescriptionKey];
        NSError *noConnectionError = [NSError errorWithDomain:NSCocoaErrorDomain
														 code:kCFURLErrorNotConnectedToInternet
													 userInfo:userInfo];
        [self handleError:noConnectionError];
    }
	else
	{
        // otherwise handle the error generically
        [self handleError:error];
    }
    
    self.appListFeedConnection = nil;   // release our connection
}

// -------------------------------------------------------------------------------
//	connectionDidFinishLoading:connection
// -------------------------------------------------------------------------------
- (void)connectionDidFinishLoading:(NSURLConnection *)connection
{
    self.appListFeedConnection = nil;   // release our connection
    
    [UIApplication sharedApplication].networkActivityIndicatorVisible = NO;   
    
    // create the queue to run our ParseOperation
    self.queue = [[[NSOperationQueue alloc] init] autorelease];
    
    // create an ParseOperation (NSOperation subclass) to parse the RSS feed data so that the UI is not blocked
    // "ownership of appListData has been transferred to the parse operation and should no longer be
    // referenced in this thread.
    //
    ParseOperation *parser = [[ParseOperation alloc] initWithData:appListData 
                                                completionHandler:^(NSArray *appList) {
        dispatch_async(dispatch_get_main_queue(), ^{
            
            [self handleLoadedApps:appList];
            
        });
        
        self.queue = nil;   // we are finished with the queue and our ParseOperation
    }];
    
    parser.errorHandler = ^(NSError *parseError) {
        dispatch_async(dispatch_get_main_queue(), ^{
            
            [self handleError:parseError];
            
        });
    };
    
    [queue addOperation:parser]; // this will start the "ParseOperation"
    
    [parser release];
    
    // ownership of appListData has been transferred to the parse operation
    // and should no longer be referenced in this thread
    self.appListData = nil;
}

// -------------------------------------------------------------------------------
//	dealloc
// -------------------------------------------------------------------------------
- (void)dealloc
{
	[navigationController release];
    [appRecords release];
    
	[rootViewController release];
    [appListFeedConnection release];
    [appListData release];
    
    [window release];
    
	[super dealloc];
}

@end

File: UIKit and Animations/WWDC 2011/SampleCodes/LazyTableImages/LazyTable-Swift/LazyTable/LazyTableAppDelegate.swift
/*
  LazyTableAppDelegate.swift
  LazyTable

  Created by Seyed Samad Gholamzadeh on 6/25/18.
  Copyright © 2018 Seyed Samad Gholamzadeh. All rights reserved.


Abstract: Application delegate for the LazyTableImage sample.
It also downloads in the background the "Top paid iPhone apps" RSS feed using URLSession.
*/


import UIKit

// This framework was imported so we could use the kCFURLErrorNotConnectedToInternet error code.
import CFNetwork

let TopPaidAppsFeed =
"http://phobos.apple.com/WebObjects/MZStoreServices.woa/ws/RSS/toppaidapplications/limit=75/xml";


@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {

	var window: UIWindow?
	var rootViewController: RootViewController!
	var appRecords: [AppRecord]!
	var queue: OperationQueue!
	
	// -------------------------------------------------------------------------------
	//	applicationDidFinishLaunching:application
	// -------------------------------------------------------------------------------
	func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {

		let nav = window!.rootViewController as! UINavigationController
		self.rootViewController = nav.topViewController as! RootViewController
		self.appRecords = []
		// Initialize the array of app records and pass a reference to that list to our root view controller
		
		let urlRequest = URLRequest(url: URL(string: TopPaidAppsFeed)!)
		
		// Test the validity of the connection object. The most likely reason for the connection object
		// to be nil is a malformed URL, which is a programmatic error easily detected during development
		// If the URL is more dynamic, then you should implement a more flexible validation technique, and
		// be able to both recover from errors and communicate problems to the user in an unobtrusive manner.
		//
		let task = URLSession.shared.dataTask(with: urlRequest) { (data, responce, error) in
			
			guard error == nil else {
				self.handleError(error!)
				return
			}
			
			
			// create the queue to run our ParseOperation

			self.queue = OperationQueue()
			
			// create an ParseOperation (NSOperation subclass) to parse the RSS feed data so that the UI is not blocked
			// "ownership of appListData has been transferred to the parse operation and should no longer be
			// referenced in this thread.
			//
			let parser = ParseOperation(data: data!, completionHandler: { (appList) in
				DispatchQueue.main.async {
					self.handleLoadedApps(appList)
				}
				self.queue = nil
			})
			
			parser.errorHandler = { error in
				DispatchQueue.main.async {
					self.handleError(error)
				}
			}
			
			self.queue.addOperation(parser)  // this will start the "ParseOperation"

		}
		
		UIApplication.shared.isNetworkActivityIndicatorVisible = true
		task.resume()
		
		return true
	}
	
	// -------------------------------------------------------------------------------
	//	handleLoadedApps:notif
	// -------------------------------------------------------------------------------
	func handleLoadedApps(_ loadedApps: [AppRecord]) {
		self.appRecords.append(contentsOf: loadedApps)
		rootViewController.entries = self.appRecords
		// tell our table view to reload its data, now that parsing has completed
		self.rootViewController.tableView.reloadData()
	}
	
	// -------------------------------------------------------------------------------
	//	handleError:error
	// -------------------------------------------------------------------------------
	func handleError(_ error: Error) {
		let errorMessage = error.localizedDescription
		let alertView = UIAlertController(title: "Cannot Show Top Paid Apps", message: errorMessage, preferredStyle: .alert)
		alertView.addAction(UIAlertAction(title: "OK", style: .cancel, handler: nil))
		self.rootViewController.present(alertView, animated: true, completion: nil)
	}
	
}

